File: 153 Find Minimum in Rotated Sorted Array.c
#include <stdio.h>
/*
153 Find Minimum in Rotated Sorted Array
https://oj.leetcode.com/problems/find-minimum-in-rotated-sorted-array/
*/
#define MIN(a,b) (a)<(b)?(a):(b)
int search(int* arr, int size)
{
    int lt,rt,mid,rightmost;

    lt = 0;
    rt = size - 1;
    rightmost = arr[rt];

    while(lt <= rt)
    {
        if(lt == rt)
            return arr[lt];
        if(lt + 1 == rt)
            return MIN(arr[lt],arr[rt]);
            
        mid = (lt + rt) >> 1;
        if(arr[mid-1] > arr[mid])
            return arr[mid];
       
        /*
        Notice the reason to use right most rather than
        left most. The array can have no turn.
        */
        if(arr[mid] > rightmost)
        {
            lt = mid + 1; 
        }
        else
        {   
            rt = mid - 1;
        }
    }
}
int main()
{
    int arr[] = {3,1,1};
    int arr[] = {1,1,3};
    int r = search(arr,sizeof(arr)/sizeof(int));
    printf("%d\n",r);
    return 0;
}

File: 153 Find Minimum in Rotated Sorted Array.java
/*
153. Find Minimum in Rotated Sorted Array
https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

Suppose an array sorted in ascending order is rotated at some pivot unknown to you beforehand.

(i.e.,  [0,1,2,4,5,6,7] might become  [4,5,6,7,0,1,2]).

Find the minimum element.

You may assume no duplicate exists in the array.

Example 1:

Input: [3,4,5,1,2] 
Output: 1

Example 2:

Input: [4,5,6,7,0,1,2]
Output: 0

Example 3:
Input: [1,2,3]
Output: 1

Example 4:
Input: [0]
Output: 0
*/
class Solution
{
    public int findMin(int[] arr)
    {   
        int left = 0;
        int right = arr.length - 1;
        int lastVal = arr[right];
        
        while (left <= right)
        {         
            // Bottom cases
            if (left == right)
            {
                return arr[left];
            }
            if (left + 1 == right)
            {
                return Math.min(arr[left],arr[right]);
            }
            
            int mid = (left + right) >> 1;
            
            // mid - 1 is guaranteed inbound
            if (arr[mid-1] > arr[mid])
            {
                return arr[mid];
            }
            
            if (arr[mid] > lastVal)
            {
                left = mid + 1;
            }
            else
            {
                right = mid - 1;
            }
        }
        return 0; // Not reachable
    }
}
